python/ray/serve/http_proxy.py

```diff
@@ -110,18 +112,19 @@ async def __call__(self, scope, receive, send):
         http_body_bytes = await self.receive_http_body(scope, receive, send)
 
         headers = {k.decode(): v.decode() for k, v in scope["headers"]}
-        request_metadata = RequestMetadata(
-            get_random_letters(10),  # Used for debugging.
+
+        handle = RayServeHandle(
+            self.controller_handle,
             endpoint_name,
-            TaskContext.Web,
-            http_method=scope["method"].upper(),
-            call_method=headers.get("X-SERVE-CALL-METHOD".lower(), "__call__"),
+            sync=True,
+            method_name=headers.get("X-SERVE-CALL-METHOD".lower(), None),
             shard_key=headers.get("X-SERVE-SHARD-KEY".lower(), None),
-        )
+            http_method=scope["method"].upper(),
+            http_headers=headers)
+
+        request = build_flask_request(scope, io.BytesIO(http_body_bytes))
 
-        ref = await self.router.assign_request(request_metadata, scope,
-                                               http_body_bytes)
-        result = await ref
+        result = await handle.remote(request)
```

Comment:
We should be caching the handle in the HTTP proxy instead of constructing a new one for each request. This is important for two reasons: (1) constructing the handle has some static overhead that we're introducing for each request and (2) the handle has some internal logic to maintain how many requests are outstanding to each replica, which is important for batching.